index.js

```diff
@@ -73,37 +67,103 @@ WebNode.prototype = {
 
     return null;
   },
-  setText(text) {
-    perform(this._webView, SET_ELEMENT_TEXT_SCRIPT, {
+  async setText(text) {
+    return perform(this._webView, SET_ELEMENT_TEXT_SCRIPT, {
       ref: this.ref,
       text: text
     });
   },
-  click() {
-    perform(this._webView, CLICK_ELEMENT_SCRIPT, {
+  async click() {
+    return perform(this._webView, CLICK_ELEMENT_SCRIPT, {
       ref: this.ref
     });
   },
-  tap() {
-    perform(this._webView, TAP_ELEMENT_SCRIPT, {
+  async tap() {
+    return perform(this._webView, TAP_ELEMENT_SCRIPT, {
       ref: this.ref
     });
   },
-  getRect() {
-    return perform(this._webView, GET_ELEMENT_RECT_SCRIPT, {
+  async getRect() {
+    const result = await perform(this._webView, GET_ELEMENT_RECT_SCRIPT, {
       ref: this.ref
-    }).rect;
+    });
+    return result.rect;
   },
-  isVisible() {
-    return perform(this._webView, IS_ELEMENT_VISIBLE_SCRIPT, {
+  async isVisible() {
+    const result = await perform(this._webView, IS_ELEMENT_VISIBLE_SCRIPT, {
       ref: this.ref
-    }).visible;
+    });
+    return result.visible;
   }
 };
 
 function perform(webView, script, params) {
-  const rawResult = webView.stringByEvaluatingJavaScriptFromString_('JSON.stringify((' + script + ').call(this, ' + JSON.stringify(params) + '));');
-  const result = JSON.parse(rawResult.toString());
+  const paramsString = (params !== undefined) ? `, ${JSON.stringify(params)}` : '';
+  const scriptString = `JSON.stringify((${script}).call(this${paramsString}));`;
+
+  if ('stringByEvaluatingJavaScriptFromString_' in webView) {
+    // UIWebView
+    return new Promise((resolve, reject) => {
+      if (isMainThread()) {
+        evaluateJavascript();
+      } else {
+        ObjC.schedule(ObjC.mainQueue, () => {
+          evaluateJavascript();
+        });
+      }
+
+      function evaluateJavascript () {
+        const rawResult = webView.stringByEvaluatingJavaScriptFromString_(scriptString);
+        try {
+          const result = parseResult(rawResult);
+          resolve(result);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    });
+  } else if ('evaluateJavaScript_completionHandler_' in webView) {
+    // WKWebView
+    return new Promise((resolve, reject) => {
+      const completionHandler = new ObjC.Block({
+        retType: 'void',
+        argTypes: ['object', 'pointer'],
+        implementation: function (rawResult, error) {
+          if (!error.isNull()) {
+            const err = new ObjC.Object(error);
+            reject(new Error(err.toString()));
```

Comment:
Should probably move the `try` block into an `else`, or return early and move the `pendingBlocks.delete()` to the top.

Reply:
oops 😊